File: day-18/part-2.py
#!/usr/bin/env python3

import json
import math
import re
import sys


def _split(number, left_split=False):
    new_number = []
    for element in number:
        if isinstance(element, int) and not left_split and element >= 10:
            new_number.append([element // 2, math.ceil(element / 2)])
            left_split = True
        elif isinstance(element, list):
            new_element, left_split = _split(element, left_split)
            new_number.append(new_element)
        else:
            new_number.append(element)
    return (new_number, left_split)


def split(number):
    return _split(number)[0]


def explode(number):
    nesting = 0
    number_str = str(number)
    for i, c in enumerate(number_str):
        if c == "[":
            nesting += 1
        elif c == "]":
            nesting -= 1

        if nesting == 5 and (match := re.match(r"\[(\d+), (\d+)\]", number_str[i:])):
            left_explode = int(match.group(1))
            right_explode = int(match.group(2))

            right_match = re.search(r"(\d+)", number_str[i + match.end() :])
            if right_match:
                number_str = (
                    number_str[: i + match.end() + right_match.start(1)]
                    + str(int(right_match.group(1)) + right_explode)
                    + number_str[i + match.end() + right_match.end(1) :]
                )

            number_str = number_str[:i] + "0" + number_str[i + match.end() :]

            left_match = re.search(r".*\b(\d+)", number_str[:i])
            if left_match:
                number_str = (
                    number_str[: left_match.start(1)]
                    + str(int(left_match.group(1)) + left_explode)
                    + number_str[left_match.end(1) :]
                )

            return True, json.loads(number_str)
    return False, number


def reduce(number):
    while True:
        changed, number = explode(number)
        if changed:
            continue

        if (split_number := split(number)) != number:
            number = split_number
            continue

        return number


def magnitude(entry):
    if isinstance(entry, int):
        return entry
    else:
        return 3 * magnitude(entry[0]) + 2 * magnitude(entry[1])


if __name__ == "__main__":
    numbers = [json.loads(line.strip()) for line in open(sys.argv[1]).readlines()]
    print(max(magnitude(reduce([a, b])) for a in numbers for b in numbers if a != b))

File: day-18/part-2.js
#!/usr/bin/env node

const fs = require("fs").promises;

const _split = (number, leftSplit = false) => {
  const newNumber = [];

  number.forEach((element) => {
    if (Number.isInteger(element) && !leftSplit && element >= 10) {
      newNumber.push([Math.floor(element / 2), Math.ceil(element / 2)]);
      leftSplit = true;
    } else if (Array.isArray(element)) {
      [newElement, leftSplit] = _split(element, leftSplit);
      newNumber.push(newElement);
    } else {
      newNumber.push(element);
    }
  });

  return [newNumber, leftSplit];
};

const split = (number) => {
  const [new_number, leftSplit] = _split(number);
  return [leftSplit, new_number];
};

const explode = (number) => {
  let nesting = 0;
  let numberStr = JSON.stringify(number);

  for (let i = 0, c = ""; (c = numberStr.charAt(i)); i++) {
    if (c === "[") {
      nesting++;
    } else if (c === "]") {
      nesting--;
    }

    if (
      nesting == 5 &&
      (match = numberStr.substring(i).match(/^\[(?<left>\d+),(?<right>\d+)\]/))
    ) {
      const { left, right } = match.groups;
      const matchEnd = match.index + 3 + left.length + right.length;

      if (
        (rightMatch = numberStr.substring(i + matchEnd).match(/(?<num>\d+)/))
      ) {
        const { num } = rightMatch.groups;
        numberStr =
          numberStr.substring(0, i + matchEnd + rightMatch.index) +
          (Number.parseInt(num) + Number.parseInt(right)) +
          numberStr.substring(i + matchEnd + rightMatch.index + num.length);
      }

      numberStr = `${numberStr.substring(0, i)}0${numberStr.substring(
        i + matchEnd
      )}`;

      if (
        (leftMatch = numberStr
          .substring(0, i)
          .match(/(?<lead>.*\b)(?<num>\d+)/))
      ) {
        const { lead, num } = leftMatch.groups;
        numberStr =
          numberStr.substring(0, leftMatch.index + lead.length) +
          (Number.parseInt(num) + Number.parseInt(left)) +
          numberStr.substring(leftMatch.index + lead.length + num.length);
      }

      return [true, JSON.parse(numberStr)];
    }
  }

  return [false, number];
};

const reduce = (number) => {
  while (true) {
    [changed, number] = explode(number);
    if (changed) continue;

    [changed, number] = split(number);
    if (changed) continue;

    return number;
  }
};

const magnitude = (number) => {
  if (Number.isInteger(number)) {
    return number;
  } else {
    return 3 * magnitude(number[0]) + 2 * magnitude(number[1]);
  }
};

(async () => {
  const numbers = (
    await fs.readFile(
      process.argv[process.argv.length - 1],
      {
        encoding: "utf8",
      },
      "utf8"
    )
  )
    .split("\n")
    .map((l) => JSON.parse(l));

  console.log(
    Math.max(
      ...numbers.flatMap((a) => numbers.map((b) => magnitude(reduce([a, b]))))
    )
  );
})();
